text/0000-macro-naming.md

```diff
@@ -0,0 +1,169 @@
+- Feature Name: N/A (part of other unstable features)
+- Start Date: 2016-02-11
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Naming and modularisation for macros.
+
+This RFC proposes making macros a first-class citizen in the Rust module system.
+Both macros by example (`macro_rules` macros) and procedural macros (aka syntax
+extensions) would use the same naming and modularisation scheme as other items
+in Rust.
+
+For procedural macros, this RFC could be implemented immediately or as part of a
+larger effort to reform procedural macros. For macros by example, this would be
+part of a macros 2.0 feature, the rest of which will be described in a separate
+RFC. This RFC depends on the changes to name resolution described in
+[RFC 1560](https://github.com/rust-lang/rfcs/pull/1560).
+
+# Motivation
+[motivation]: #motivation
+
+Currently, procedural macros are not modularised at all (beyond the crate
+level). Macros by example have a [custom modularisation
+scheme](https://github.com/rust-lang/rfcs/blob/master/text/0453-macro-reform.md)
+which involves modules to some extent, but relies on source ordering and
+attributes which are not used for other items. Macros cannot be imported or
+named using the usual syntax. It is confusing that macros use their own system
+for modularisation. It would be far nicer if they were a more regular feature of
+Rust in this respect.
+
+
+# Detailed design
+[design]: #detailed-design
+
+## Defining macros
+
+This RFC does not propose changes to macro definitions. It is envisaged that
+definitions of procedural macros will change, see [this blog post](http://ncameron.org/blog/macro-plans-syntax/)
+for some rough ideas. I'm assuming that procedural macros will be defined in
+some function-like way and that these functions will be defined in modules in
+their own crate (to start with).
+
+Ordering of macro definitions in the source text will no longer be significant.
+A macro may be used before it is defined, as long as it can be named. That is,
+macros follow the same rules regarding ordering as other items. E.g., this will
+work:
+
+```
+foo!();
+
+macro! foo { ... }
+```
+
```

Comment:
I don't think this is a good idea, because it won't play nice with macro-defining macros. For example, the following sort of thing should (and currently does) work:
```
  macro_rules! macro_defining { 
    ($name:ident) => { macro_rules! $name { () => (println!("Yay!");) } }
  }

  macro_defining!( say_yay );

  say_yay!();
```
...but it only works if the macros are expanded in the order they are written, and there's no reasonable static analysis that can tell us that fact.